### ISSUE TITLE
`tf.random.uniform()` always picking the same value when metal plugin is installed

### ISSUE THREAD

**Issue by yunhao-qian** [tensorflow/tensorflow#56122]
### System information

* OS: macOS Montererey 12.4
* Tensorflow installed from: `pip` commands inside a conda environment
* Tensorflow version: `tensorflow-macos` 2.8.0
* Metal plugin version: `tensorflow-metal` 0.4.0
* Python version: 3.9.12
* GPU model and memory: Apple M1 Pro with 16 GB memory
* Exact command to reproduce: see below

### Describe the problem

When the `tensorflow-metal` plugin is installed, the `tf.random.uniform()` function always picks the same value on every call. Other random functions like `tf.random.normal()` do not have this problem. Neither does this problem occur when the Metal plugin is not installed.

### Source code / logs

```text
yunhao@Yunhaos-MBP ~ % conda create --name tf-rand-test python=3.9 six=1.15                      
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/yunhao/miniforge3/envs/tf-rand-test

  added / updated specs:
    - python=3.9
    - six=1.15


The following NEW packages will be INSTALLED:

  bzip2              conda-forge/osx-arm64::bzip2-1.0.8-h3422bc3_4
  ca-certificates    conda-forge/osx-arm64::ca-certificates-2021.10.8-h4653dfc_0
  libffi             conda-forge/osx-arm64::libffi-3.4.2-h3422bc3_5
  libzlib            conda-forge/osx-arm64::libzlib-1.2.11-h90dfc92_1014
  ncurses            conda-forge/osx-arm64::ncurses-6.3-h07bb92c_1
  openssl            conda-forge/osx-arm64::openssl-3.0.3-ha287fd2_0
  pip                conda-forge/noarch::pip-22.1-pyhd8ed1ab_0
  python             conda-forge/osx-arm64::python-3.9.12-h14b404e_1_cpython
  python_abi         conda-forge/osx-arm64::python_abi-3.9-2_cp39
  readline           conda-forge/osx-arm64::readline-8.1-hedafd6a_0
  setuptools         conda-forge/osx-arm64::setuptools-62.2.0-py39h2804cbe_0
  six                conda-forge/noarch::six-1.15.0-pyh9f0ad1d_0
  sqlite             conda-forge/osx-arm64::sqlite-3.38.5-h40dfcc0_0
  tk                 conda-forge/osx-arm64::tk-8.6.12-he1e0b03_0
  tzdata             conda-forge/noarch::tzdata-2022a-h191b570_0
  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0
  xz                 conda-forge/osx-arm64::xz-5.2.5-h642e427_1
  zlib               conda-forge/osx-arm64::zlib-1.2.11-h90dfc92_1014


Proceed ([y]/n)? 

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate tf-rand-test
#
# To deactivate an active environment, use
#
#     $ conda deactivate

yunhao@Yunhaos-MBP ~ % conda activate tf-rand-test                         
(tf-rand-test) yunhao@Yunhaos-MBP ~ % conda install -c apple tensorflow-deps
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/yunhao/miniforge3/envs/tf-rand-test

  added / updated specs:
    - tensorflow-deps


The following NEW packages will be INSTALLED:

  c-ares             conda-forge/osx-arm64::c-ares-1.18.1-h3422bc3_0
  cached-property    conda-forge/noarch::cached-property-1.5.2-hd8ed1ab_1
  cached_property    conda-forge/noarch::cached_property-1.5.2-pyha770c72_1
  grpcio             conda-forge/osx-arm64::grpcio-1.46.1-py39h365d37b_0
  h5py               conda-forge/osx-arm64::h5py-3.6.0-nompi_py39hd982b79_100
  hdf5               conda-forge/osx-arm64::hdf5-1.12.1-nompi_hd9dbc9e_104
  krb5               conda-forge/osx-arm64::krb5-1.19.3-he492e65_0
  libblas            conda-forge/osx-arm64::libblas-3.9.0-14_osxarm64_openblas
  libcblas           conda-forge/osx-arm64::libcblas-3.9.0-14_osxarm64_openblas
  libcurl            conda-forge/osx-arm64::libcurl-7.83.1-h7965298_0
  libcxx             conda-forge/osx-arm64::libcxx-14.0.3-h6a5c8ee_0
  libedit            conda-forge/osx-arm64::libedit-3.1.20191231-hc8eb9b7_2
  libev              conda-forge/osx-arm64::libev-4.33-h642e427_1
  libgfortran        conda-forge/osx-arm64::libgfortran-5.0.0.dev0-11_0_1_hf114ba7_23
  libgfortran5       conda-forge/osx-arm64::libgfortran5-11.0.1.dev0-hf114ba7_23
  liblapack          conda-forge/osx-arm64::liblapack-3.9.0-14_osxarm64_openblas
  libnghttp2         conda-forge/osx-arm64::libnghttp2-1.47.0-hf30690b_0
  libopenblas        conda-forge/osx-arm64::libopenblas-0.3.20-openmp_h2209c59_0
  libssh2            conda-forge/osx-arm64::libssh2-1.10.0-h7a5bd25_2
  llvm-openmp        conda-forge/osx-arm64::llvm-openmp-14.0.3-hd125106_0
  numpy              conda-forge/osx-arm64::numpy-1.21.6-py39h690d673_0
  tensorflow-deps    apple/osx-arm64::tensorflow-deps-2.8.0-0


Proceed ([y]/n)? 

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
(tf-rand-test) yunhao@Yunhaos-MBP ~ % python -m pip install tensorflow-macos                      
Collecting tensorflow-macos
  Using cached tensorflow_macos-2.8.0-cp39-cp39-macosx_11_0_arm64.whl (190.1 MB)
Collecting tensorboard<2.9,>=2.8
  Using cached tensorboard-2.8.0-py3-none-any.whl (5.8 MB)
Collecting absl-py>=0.4.0
  Using cached absl_py-1.0.0-py3-none-any.whl (126 kB)
Collecting libclang>=9.0.1
  Using cached libclang-14.0.1-py2.py3-none-macosx_11_0_arm64.whl (11.8 MB)
Collecting termcolor>=1.1.0
  Using cached termcolor-1.1.0-py3-none-any.whl
Collecting protobuf>=3.9.2
  Using cached protobuf-3.20.1-py2.py3-none-any.whl (162 kB)
Collecting opt-einsum>=2.3.2
  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)
Collecting keras-preprocessing>=1.1.1
  Using cached Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)
Collecting flatbuffers>=1.12
  Using cached flatbuffers-2.0-py2.py3-none-any.whl (26 kB)
Requirement already satisfied: six>=1.12.0 in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from tensorflow-macos) (1.15.0)
Collecting keras<2.9,>=2.8.0rc0
  Using cached keras-2.8.0-py2.py3-none-any.whl (1.4 MB)
Requirement already satisfied: setuptools in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from tensorflow-macos) (62.2.0)
Collecting astunparse>=1.6.0
  Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
Collecting wrapt>=1.11.0
  Using cached wrapt-1.14.1-cp39-cp39-macosx_11_0_arm64.whl (35 kB)
Collecting gast>=0.2.1
  Using cached gast-0.5.3-py3-none-any.whl (19 kB)
Collecting google-pasta>=0.1.1
  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)
Collecting typing-extensions>=3.6.6
  Using cached typing_extensions-4.2.0-py3-none-any.whl (24 kB)
Requirement already satisfied: numpy>=1.20 in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from tensorflow-macos) (1.21.6)
Requirement already satisfied: grpcio<2.0,>=1.24.3 in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from tensorflow-macos) (1.46.1)
Collecting tf-estimator-nightly==2.8.0.dev2021122109
  Using cached tf_estimator_nightly-2.8.0.dev2021122109-py2.py3-none-any.whl (462 kB)
Requirement already satisfied: h5py>=2.9.0 in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from tensorflow-macos) (3.6.0)
Requirement already satisfied: wheel<1.0,>=0.23.0 in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from astunparse>=1.6.0->tensorflow-macos) (0.37.1)
Collecting google-auth<3,>=1.6.3
  Using cached google_auth-2.6.6-py2.py3-none-any.whl (156 kB)
Collecting requests<3,>=2.21.0
  Using cached requests-2.27.1-py2.py3-none-any.whl (63 kB)
Collecting tensorboard-plugin-wit>=1.6.0
  Using cached tensorboard_plugin_wit-1.8.1-py3-none-any.whl (781 kB)
Collecting werkzeug>=0.11.15
  Using cached Werkzeug-2.1.2-py3-none-any.whl (224 kB)
Collecting markdown>=2.6.8
  Using cached Markdown-3.3.7-py3-none-any.whl (97 kB)
Collecting tensorboard-data-server<0.7.0,>=0.6.0
  Using cached tensorboard_data_server-0.6.1-py3-none-any.whl (2.4 kB)
Collecting google-auth-oauthlib<0.5,>=0.4.1
  Using cached google_auth_oauthlib-0.4.6-py2.py3-none-any.whl (18 kB)
Collecting pyasn1-modules>=0.2.1
  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
Collecting cachetools<6.0,>=2.0.0
  Using cached cachetools-5.1.0-py3-none-any.whl (9.2 kB)
Collecting rsa<5,>=3.1.4
  Using cached rsa-4.8-py3-none-any.whl (39 kB)
Collecting requests-oauthlib>=0.7.0
  Using cached requests_oauthlib-1.3.1-py2.py3-none-any.whl (23 kB)
Collecting importlib-metadata>=4.4
  Using cached importlib_metadata-4.11.3-py3-none-any.whl (18 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.3-py3-none-any.whl (61 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.9-py2.py3-none-any.whl (138 kB)
Collecting charset-normalizer~=2.0.0
  Using cached charset_normalizer-2.0.12-py3-none-any.whl (39 kB)
Collecting zipp>=0.5
  Using cached zipp-3.8.0-py3-none-any.whl (5.4 kB)
Collecting pyasn1<0.5.0,>=0.4.6
  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
Collecting oauthlib>=3.0.0
  Using cached oauthlib-3.2.0-py3-none-any.whl (151 kB)
Installing collected packages: tf-estimator-nightly, termcolor, tensorboard-plugin-wit, pyasn1, libclang, keras, flatbuffers, certifi, zipp, wrapt, werkzeug, urllib3, typing-extensions, tensorboard-data-server, rsa, pyasn1-modules, protobuf, opt-einsum, oauthlib, keras-preprocessing, idna, google-pasta, gast, charset-normalizer, cachetools, astunparse, absl-py, requests, importlib-metadata, google-auth, requests-oauthlib, markdown, google-auth-oauthlib, tensorboard, tensorflow-macos
Successfully installed absl-py-1.0.0 astunparse-1.6.3 cachetools-5.1.0 certifi-2021.10.8 charset-normalizer-2.0.12 flatbuffers-2.0 gast-0.5.3 google-auth-2.6.6 google-auth-oauthlib-0.4.6 google-pasta-0.2.0 idna-3.3 importlib-metadata-4.11.3 keras-2.8.0 keras-preprocessing-1.1.2 libclang-14.0.1 markdown-3.3.7 oauthlib-3.2.0 opt-einsum-3.3.0 protobuf-3.20.1 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.27.1 requests-oauthlib-1.3.1 rsa-4.8 tensorboard-2.8.0 tensorboard-data-server-0.6.1 tensorboard-plugin-wit-1.8.1 tensorflow-macos-2.8.0 termcolor-1.1.0 tf-estimator-nightly-2.8.0.dev2021122109 typing-extensions-4.2.0 urllib3-1.26.9 werkzeug-2.1.2 wrapt-1.14.1 zipp-3.8.0
(tf-rand-test) yunhao@Yunhaos-MBP ~ % python                                
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:25:14) 
[Clang 12.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> for _ in range(10):
...     print(tf.random.uniform([]))
... 
tf.Tensor(0.43182516, shape=(), dtype=float32)
tf.Tensor(0.73127687, shape=(), dtype=float32)
tf.Tensor(0.8234819, shape=(), dtype=float32)
tf.Tensor(0.7472857, shape=(), dtype=float32)
tf.Tensor(0.06519365, shape=(), dtype=float32)
tf.Tensor(0.46567404, shape=(), dtype=float32)
tf.Tensor(0.081846, shape=(), dtype=float32)
tf.Tensor(0.26130438, shape=(), dtype=float32)
tf.Tensor(0.53803635, shape=(), dtype=float32)
tf.Tensor(0.98602235, shape=(), dtype=float32)
>>> exit()
(tf-rand-test) yunhao@Yunhaos-MBP ~ % python -m pip install tensorflow-metal                      
Collecting tensorflow-metal
  Using cached tensorflow_metal-0.4.0-cp39-cp39-macosx_11_0_arm64.whl (1.2 MB)
Requirement already satisfied: wheel~=0.35 in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from tensorflow-metal) (0.37.1)
Requirement already satisfied: six~=1.15.0 in ./miniforge3/envs/tf-rand-test/lib/python3.9/site-packages (from tensorflow-metal) (1.15.0)
Installing collected packages: tensorflow-metal
Successfully installed tensorflow-metal-0.4.0
(tf-rand-test) yunhao@Yunhaos-MBP ~ % python                                
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:25:14) 
[Clang 12.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> for _ in range(10):
...     print(tf.random.uniform([]))
... 
Metal device set to: Apple M1 Pro

systemMemory: 16.00 GB
maxCacheSize: 5.33 GB

2022-05-16 15:21:55.560387: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2022-05-16 15:21:55.560711: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:271] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
>>> for _ in range(10):
...     print(tf.random.uniform([]))
... 
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
tf.Tensor(0.9149647, shape=(), dtype=float32)
>>> exit()
(tf-rand-test) yunhao@Yunhaos-MBP ~ % 
```

**Comment by sachinprasadhs** [comment]
It is because the random.uniform in apple metal is not able to initialize the op seed or global seed, when both of these seeds values are set to None, you'll get the same number generated. 
To avoid this you can enable `tf.config.experimental.enable_op_determinism()` and set the `random_seed` to generate different results. Refer the below code.

```
import tensorflow as tf
tf.keras.utils.set_random_seed(1)
tf.config.experimental.enable_op_determinism()

for _ in range(10):
  print(tf.random.uniform([]))
```

**Comment by yunhao-qian** [comment]
Thank you! 😊 After the global seed is set, `tf.random.uniform()` behaves as expected.

When I read the documentation of `tf.random.set_seed()`, I found the following sentence which complements your explanation:

> If neither the global seed nor the operation seed is set: A randomly picked seed is used for this op.

So here, how the seed is "randomly picked" is not clearly specified. Should I understand it as: since Metal is unable to initialize the global seed or the operation seed, the current behaviour (without seeding) is an expected "sharp edge" instead of a bug? This can be very surprising for many people like me, so maybe it should be documented somewhere more explicitly.

Thank you again for your help. You can close this issue if you think no further action is needed.
